[SwiftUI] When using the `webViewOnScrollGeometryChange` view modifier, the affected WebView is redrawn with each scroll event

rr-codes opened the pull request:
#### d36b858cb4e83f4a7be9bf86ef697e3ddbdf82f4
<pre>
[SwiftUI] When using the `webViewOnScrollGeometryChange` view modifier, the affected WebView is redrawn with each scroll event
<a href="https://bugs.webkit.org/show_bug.cgi?id=288645">https://bugs.webkit.org/show_bug.cgi?id=288645</a>
<a href="https://rdar.apple.com/145688135">rdar://145688135</a>

Reviewed by Aditya Keerthi.

SwiftUI layout results in the environment being re-evaluated to see if any values differ. In the case of
`webViewOnScrollGeometryChange`, this means that a new `OnScrollGeometryChangeContext` was created each time and the environment
updated.

When the environment is updated, SwiftUI calls the `update{UI/NS}View}` function of the view representable, and redraws the view.
As a result, each time a scroll event happens, all the logic in the update function was being unnecessarily invoked, and the view
was unnecessarily redrawn.

To fix this, ensure that only ever one `OnScrollGeometryChangeContext` is in the environment. The logic the client uses in the view
modifier is encapsulated inside an internal dictionary in the context. This is needed to ensure that if the view modifier is used
multiple times, that each transform/action is stored individually instead of replaced. This dictioanry is keyed by `Namespace` so that
if either of the closures change within the _same_ view modifier application, they are simply updated instead of a new change being added.

* Source/WebKit/_WebKit_SwiftUI/CocoaWebViewAdapter.swift:
(CocoaWebViewAdapter.geometryDidChange(_:)):
* Source/WebKit/_WebKit_SwiftUI/Foundation+Extras.swift:
* Source/WebKit/_WebKit_SwiftUI/View+WebViewModifiers.swift:
(OnScrollGeometryChangeModifier.body(_:)):
(EnvironmentValues.webViewOnScrollGeometryChange): Deleted.
* Source/WebKit/_WebKit_SwiftUI/ViewModifierContexts.swift:
(Change.transform):
(Change.action):
(changes):
(apply(from:to:)):

Canonical link: <a href="https://commits.webkit.org/291350@main">https://commits.webkit.org/291350@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/134066061cd9be225ce89f97e1506e6eaaac5dae

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/92748 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/131/builds/12299 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/138/builds/1939 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/97744 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/43259 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/12577 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/20751 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/97744 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/43259 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/95750 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/9514 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/138/builds/1939 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/97744 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/9204 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/138/builds/1939 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/42587 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/79514 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/138/builds/1939 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/99765 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/19800 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/14567 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/99765 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/20051 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/138/builds/1939 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/99765 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/138/builds/1939 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/12814 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14787 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/19784 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/24960 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/19471 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/22931 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/21212 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebKit/_WebKit_SwiftUI/ViewModifierContexts.swift:
```diff
@@ -30,13 +31,23 @@ struct ContextMenuContext {
     var canReplace = true
 }
+
+struct ScrollPositionContext {
```
remove this

axiixc commented on Source/WebKit/_WebKit_SwiftUI/ViewModifierContexts.swift:
```diff
@@ -30,9 +31,15 @@ struct ContextMenuContext {
+    var action: (AnyEquatable, AnyEquatable) -> Void
+
+    init() {
+        transform = { _ in AnyEquatable()  }
```
```suggestion
        transform = { _ in AnyEquatable() }
```

axiixc commented on Source/WebKit/_WebKit_SwiftUI/Foundation+Extras.swift:
```diff
@@ -35,16 +35,55 @@ func onNextMainRunLoop(do body: @escaping @MainActor () -> Void) {
+    }
+
+    /// Creates an `AnyEquatable` that is never equal to anything.
+    init() {
```
I’m tempted to say you should call this `.nan` given that `AnyHashable() == AnyHashable()` is now `false`. Alternately… I think it’d be neat if we didn’t need this. I can see this being a future pitfall for someone.

axiixc commented on Source/WebKit/_WebKit_SwiftUI/Foundation+Extras.swift:
```diff
@@ -35,16 +35,55 @@ func onNextMainRunLoop(do body: @escaping @MainActor () -> Void) {
+
+    private let storage: Storage
+
+    var value: Any {
```
Should `value` also be `private`? Maybe call it `base` to match [`AnyHashable`’s impl](https://github.com/swiftlang/swift/blob/1576ba4448631225363f8397ff635378b78e1cd8/stdlib/public/core/AnyHashable.swift#L189)?

axiixc commented on Source/WebKit/_WebKit_SwiftUI/Foundation+Extras.swift:
```diff
@@ -35,16 +35,55 @@ func onNextMainRunLoop(do body: @escaping @MainActor () -> Void) {
     init<E: Equatable>(_ value: E) {
-        self.value = value
-        self.equals = { ($0 as! E) == value }
+        self.storage = .value(value, { other in
```
I think you could perhaps use `AnyHashable` here, if you’re willing to limit the initializer to `some Hashable` instead (technically weird, but it gets you off the hook for a lot of this stuff, and most types you’re likely to use it with should still work then).

Otherwise, you could probably just capture the meta type directly? `case value(Any, Any.Type)` (didn’t test that though).

axiixc commented on Source/WebKit/_WebKit_SwiftUI/Foundation+Extras.swift:
```diff
@@ -35,16 +35,55 @@ func onNextMainRunLoop(do body: @escaping @MainActor () -> Void) {
+}
+
+@MainActor
+func withContinuousObservation<T>(of value: @MainActor @escaping @autoclosure () -> T, execute: @MainActor @escaping (T) -> Void) {
```
This is technically extending the [`Observation` module](https://developer.apple.com/documentation/Observation).

axiixc replied:
Doesn’t matter _now_, but I wonder if you could make this generic over `any Actor` or somehow take in the isolation so isn’t `MainActor`-only.

axiixc commented on Source/WebKit/_WebKit_SwiftUI/View+WebViewModifiers.swift:
```diff
@@ -48,7 +48,7 @@ extension EnvironmentValues {
 
     @Entry
-    var webViewOnScrollGeometryChange: OnScrollGeometryChangeContext? = nil
+    var webViewOnScrollGeometryChange = OnScrollGeometryChangeContext()
```
Instead of modifying `AnyEquatable` I bet we could sink that special case into `OnScrollGeometryChangeContext` itself. Perhaps even keeping some flag or `enum` there.

axiixc commented on Source/WebKit/_WebKit_SwiftUI/ViewModifierContexts.swift:
```diff
@@ -30,9 +30,45 @@ struct ContextMenuContext {
+            action(old.base as! T, new.base as! T)
+        }
+
+        if changes[changeID] == nil {
```
I think that without some sort of `= nil` call site, you’re going to be leaking the change observers here.

rr-codes replied:
Thanks for catching this! I added a FIXME and will address in a subsequent commit.

pxlcoder reviewed (approved):
Approving so we can try out this change.

Follow-up tracked in rdar://145869526.

rr-codes committed:
[SwiftUI] When using the `webViewOnScrollGeometryChange` view modifier, the affected WebView is redrawn with each scroll event
https://bugs.webkit.org/show_bug.cgi?id=288645
rdar://145688135

Reviewed by Aditya Keerthi.

SwiftUI layout results in the environment being re-evaluated to see if any values differ. In the case of
`webViewOnScrollGeometryChange`, this means that a new `OnScrollGeometryChangeContext` was created each time and the environment
updated.

When the environment is updated, SwiftUI calls the `update{UI/NS}View}` function of the view representable, and redraws the view.
As a result, each time a scroll event happens, all the logic in the update function was being unnecessarily invoked, and the view
was unnecessarily redrawn.

To fix this, ensure that only ever one `OnScrollGeometryChangeContext` is in the environment. The logic the client uses in the view
modifier is encapsulated inside an internal dictionary in the context. This is needed to ensure that if the view modifier is used
multiple times, that each transform/action is stored individually instead of replaced. This dictioanry is keyed by `Namespace` so that
if either of the closures change within the _same_ view modifier application, they are simply updated instead of a new change being added.

* Source/WebKit/_WebKit_SwiftUI/CocoaWebViewAdapter.swift:
(CocoaWebViewAdapter.geometryDidChange(_:)):
* Source/WebKit/_WebKit_SwiftUI/Foundation+Extras.swift:
* Source/WebKit/_WebKit_SwiftUI/View+WebViewModifiers.swift:
(OnScrollGeometryChangeModifier.body(_:)):
(EnvironmentValues.webViewOnScrollGeometryChange): Deleted.
* Source/WebKit/_WebKit_SwiftUI/ViewModifierContexts.swift:
(Change.transform):
(Change.action):
(changes):
(apply(from:to:)):

Canonical link: https://commits.webkit.org/291350@main
